### ISSUE TITLE
Ignoring a single line (parse error) makes sqlfluff ignore the remainder of the file

### ISSUE THREAD

**Issue by wijnanjo** [sqlfluff/sqlfluff#6199]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I need to ignore a few lines that give me parse errors. Appending `-- noqa` to those line(s) effectively fixes the parse error but also ignores any other true errors in the remainder of the file.

### Expected Behaviour

Any line not having the `--noqa` directive should be linted and any rule violation should be marked.

### Observed Behaviour

All lines coming after the line having `--noqa` never gets marked with errors, even if they clearly are erroneous.

### How to reproduce

Below lintresult is OK even though there are multiple errors.
```bash
sqlfluff lint --dialect sparksql - <<EOF
select 1 from tbl;
echo "this breaks sqlfluff" -- noqa
select from tbl; -- should fail
EOF
All Finished 📜 🎉!
```

Commenting the middle line now results in the 2 expected errors.
```bash
sqlfluff lint --dialect sparksql - <<EOF
select 1 from tbl;
-- echo "this breaks sqlfluff" -- noqa
select from tbl; -- should fail
EOF
== [stdin] FAIL
L:   3 | P:   1 |  PRS | Line 3, Position 1: Found unparsable section: 'select'
WARNING: Parsing errors found and dialect is set to 'sparksql'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

### Dialect

sparksql

### Version

sqlfluff, version 3.1.1
Python 3.11.9

### Configuration

no custom config

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by eliza-j-g** [comment]
According to the docs 'Inline Ignoring Errors' here: https://docs.sqlfluff.com/en/stable/reference/rules.html 
Ignoring parsing errors may lead to incorrect results as SQLFluff can misinterpret the SQL being analysed.
Because the echo statement is not parseable sql, sqlfluff cannot understand it. Using correct SQL as below shows the correct errors.
```bash
sqlfluff lint --dialect sparksql - <<EOF
select 1 from tbl;
select a from b -- noqa
select from tbl; -- should fail
EOF
```

**Comment by zzztimbo** [comment]
We're stuck having to use sqlfluff 2.3.2 because of this issue. Are there any plans to address this bug?

**Comment by peterbud** [comment]
This is working with the latest `sqlfluff` as expected.

This input:
```sql
select 1 from tbl;
-- echo "this breaks sqlfluff" -- noqa
select from tbl; -- should fail
```

is properly linted in the lines after `-- noqa`:

```bash
sqlfluff lint test.sql -d sparksql
== [test.sql] FAIL                                                                                              
L:   3 | P:   1 |  PRS | Line 3, Position 1: Found unparsable section: 'select'
L:   3 | P:  18 | LT12 | Files must end with a single trailing newline.
                       | [layout.end_of_file]
WARNING: Parsing errors found and dialect is set to 'sparksql'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

Let me know if we can close the issue.

**Comment by wijnanjo** [comment]
This is not fixed:

This input:
```sql
select 1 from tbl;
-- echo "this breaks sqlfluff" -- noqa
SElect 1 from tbl; -- should fail

```
returns correct lint errors:
```bash
== [proj/tmp/sql/test.sql] FAIL                                                                                              
L:   3 | P:   1 | CP01 | Keywords must be consistently lower case.
                       | [capitalisation.keywords]
```

**However**, this input returns no lint errors at all
```sql
select 1 from tbl;
echo "this breaks sqlfluff" -- noqa
SElect 1 from tbl; -- should fail

```

It seems kind of duplicate as [this one: ](https://github.com/sqlfluff/sqlfluff/issues/4401)

Let's make it realistic. I'm using _flink_ dialect which can't parse UDFs (`create function...`) whereas _sparksql_ can. So I'd like to ignore that single line and have the rest of the script linted:

```sql
create function ARRAY_AGGR as 'org.my.code.udf.ArraAgg'; -- noqa
SElect 1 from tbl;

```
No lint errors with `-d flink` even tough the last line is malformed and should return a lint error.

Linting with `-d sparksql` and without `noqa` gives the expected lint error:
```sql
create function ARRAY_AGGR as 'org.my.code.udf.ArraAgg';
SElect 1 from tbl;

```
```bash
L:   2 | P:   1 | CP01 | Keywords must be consistently lower case.
                       | [capitalisation.keywords]
```

**Comment by peterbud** [comment]
Hi @wijnanjo thanks for explaining it - now I understand your issue better.

Few things:
* Recovery after a parsing (PRS) error is not an easy feature, it's dialect specific whether you can do a proper recovery. However handling suppressed (`--noqa`) lines might be an easier solution
* Your ultimate solution would be however NOT to suppress unparsable flink grammar elements (or trying to parse as sparksql), but rather add the appropriate `create function` grammar to flink dialect as you did with #7623

Nevertheless I'll look into the recovery option as well.

**Comment by peterbud** [comment]
Hi @wijnanjo!

Thanks again for the detailed report and the reproduction steps. We spent some time investigating this and even prototyped a fix in a PR #7624, but after some internal discussion, we’ve decided not to move forward with that specific implementation.

**The Reasoning:**
The proposed fix added significant complexity to the `noqa` logic, which we felt made the recovery mechanism too fragile. Additionally, the fix only covered cases where the parse error was on the same line as the directive, leaving many multi-line scenarios still broken.

Since we won't be pursuing the current PR, I'm going to close this issue for now. We’ve noted the requirement for better recovery, and we’ll look into the "sync point" strategy for a future architectural update. Thanks for helping us kick the tires on this!
